Add testing documentation (#766)
CrowdStrike/glide-core: CONTRIBUTING.md

describe('Checkbox Basics', () => {});
```

> ##### `*.test.aria.ts`

Author: [Rendered](https://github.com/CrowdStrike/glide-core/blob/d9c82782995347c58f9a9663b9774a5c459d6b84/CONTRIBUTING.md#testing)

Assertions against a component's accessibility tree.

##### `*.test.basics.ts`

A bit of a misnomer, but we've yet to think of a better name.
These are assertions against a component in its initial state—even if it's not its default slot.
Think of attributes and slots provided on first render.

##### `*.test.events.ts`

Tests that exclusively assert that a certain event has been dispatched.

Don't forget to assert the event's `composed` and `bubbles` properties and, if appropriate, the event's `target`.
It's also good to assert, using `sinon`, that an event was only dispatchd once.

##### `*.test.focus.ts`

Tests that exclusively assert that an element is focused after some interaction.
Each test should mainly be asserting against `document.activeElement`.

##### `*.test.interactions.ts`

A dumping ground for most interaction assertions.
An interaction can be a simulated user interaction such as a click.
Or it can be a developer interaction—such as programmatically setting an attribute or changing a slot after initial render.

This suite can get pretty large.

> So we break down specific types of interactions into two sub-suites: `*.test.events.ts` and `*.test.focus.ts`.

Author: It would make for a long file name. But `*.test.interactions.events.ts` and `*.test.interactions.focus.ts` would probably make things clearer. What do you think?

Reviewer: I'd be up for it.

##### `*.test.visuals.ts`

Tests that assert against a baseline screenshot of a component in some state.

#### Writing ARIA tests

An ARIA test is just an assertion against a component's [accessibility tree](https://developer.mozilla.org/en-US/docs/Glossary/Accessibility_tree) in serialized form (YAML).
The resultant YAML is called a "snapshot".
We [commit](https://github.com/CrowdStrike/glide-core/tree/main/src/aria-snapshots) snapshots for every test and use those as the baseline for assertions.

> ARIA tests don't supplant manual VoiceOver testing.

Author: ```suggestion
ARIA tests don't supplant manual screenreader testing.
```

> But they let us do much less of it, and they help us catch things easily missed with VoiceOver.

Author: ```suggestion
But they let us do much less of it, and they help us catch things easily missed with a screenreader.
```

And artifacts from our latest publish are [deployed](https://github.com/CrowdStrike/glide-core/blob/main/.github/workflows/on-merge-branch-changeset-release-main.yml#L95) to the top level of the bucket.
It's a simple setup.
But it means we can't have branch names that conflict and thus overwrite top-level directories that contain artifacts.

> ### How does ARIA snapshot testing fit into overall accessibility testing?

Author: [Rendered](https://github.com/CrowdStrike/glide-core/blob/d9c82782995347c58f9a9663b9774a5c459d6b84/CONTRIBUTING.md#how-does-aria-snapshot-testing-fit-into-overall-accessibility-testing)

We use Axe to test accessibility.
But Axe only tells us that a component's accessibility tree doesn't contain any obvious violations and that the accessibility tree is valid.
What Axe doesn't tell us is whether the accessibility tree is correct.

Dropdown, for example, when open, should show up as expanded in the accessibility tree.

> But, if it doesn't, Axe won't throw because Dropdown's accessibility tree is valid when Dropdown is open but not expanded in the tree.

Author: ```suggestion
But, if it doesn't, Axe won't throw because Dropdown's accessibility tree is valid when Dropdown is open even though it's not expanded in the tree.
```

Tracking a component's accessibility tree from commit to commit helps with issues like that and more.
And it gives us confidence when making changes to a component.
If a component's accessibility tree remains unchanged after the component is changed, we can more or less be sure that the component is just as accessible as it was prior to the change.

That said, ARIA tests do not obviate manual testing with screenreaders.
Because there's no guarantee that what's in the accessibility tree is interpreted correctly by screenreaders or that a screenreader hasn't had a regression.

> But it gets us about halfway there, and it greatly reduces continual and time-consuming VoiceOver passes that we'd otherwise have to do.

Author: ```suggestion
But it gets us about halfway there, and it greatly reduces continual and time-consuming screenreader passes that we'd otherwise have to do.
```

Reviewer: Don't forget the `.` to the end of your suggestion!

But comparing screenshots generated locally (the baseline) with ones generated in CI, by a branch, is famously difficult because there are invariably differences across operating systems—especially when it comes to font rendering.
And there's no guarantee that a developer's operating system is the same as CI because developers use a variety of them.

Services like Chromatic and Percy are, in part, a solution to this problem.
But, rather than rely on a third-party service, we've chosen to solve the problem by only generating screenshots in CI.
When a PR is merged, we send the new and updated screenshots from your branch to Cloudflare R2.
When you open your next PR, those baseline screenshots are retrieved from R2 and compared against the ones generated by your branch.
We then add a link to your PR to the test report that shows the differences between the baseline screenshots and those of your branch.

### Why three separate test runners?

In short, because each is incomplete or imperfect but the best at what it does well.
This situation is far from ideal.
So we're closely watching all three with the hope that one (likely Playwright) will subsume the others.

> ### Playwright

Reviewer 2: Since this falls under "Why three separate test runners?"?

```suggestion
#### Playwright
```

End-to-end testing doesn't apply to a design system.
But visual tests do and Playwright runs them well.
It also supports ARIA testing.

> So we use it for it too.

Reviewer 2: nit:

```suggestion
So we use it for that too.
```

Its [component testing](https://playwright.dev/docs/test-components), on the other hand, is still experimental and doesn't support code coverage.
Similar for unit testing.
Playwright can be made to work for unit tests.
But it doesn't support code coverage out of the box, and its watch mode is behind a flag and only watches for changes to test files.

> ### Vitest

Reviewer 2: ```suggestion
#### Vitest
```

We use Vitest for our unit tests, which are limited to our custom ESLint and Stylelint rules.

> Vitest doesn't support visual testing, however, and it's [Browser Mode](https://vitest.dev/guide/browser), which we could use for component tests, is still experimental and buggy.

Reviewer 2: ```suggestion
Vitest doesn't support visual testing, however, and its [Browser Mode](https://vitest.dev/guide/browser), which we could use for component tests, is still experimental and buggy.
```

> ### Web Test Runner

Reviewer 2: ```suggestion
#### Web Test Runner
```
